### ISSUE TITLE
Parallel started Containers (by docker-compose v2) sometimes not registered in all configured networks

### ISSUE THREAD

**Issue by Persi** [docker/for-mac#6918]
### Description

Since upgrade from Docker for Mac 4.20.1 to 4.21.1 we sometimes experience wrong container setups which cause networking issues. The reason seems to be that sometimes a few containers (every time different ones!) are not correctly registered in all of their defined networks.

Because of that e.g. our backends cannot find it's own database container (both use the same internal network defined in one docker-compose file).

I could reproduce this problem sometimes  and collected the relevant container infos in the files below.

For clarity and safety reasons i've reduced the container to the relevant minimum.
Working container setup with correct network settings: 

```
[
    {
        "Id": "62cc7a1eccd889e46df0d32e2ef60127e1905b9a7936a3ff2817711a3d209f52",
        "Created": "2023-07-14T06:33:23.40993184Z",
        "Path": "java",
        "Args": [
        ...
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 57927,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2023-07-14T06:33:24.125459174Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:1a7fbe9d41a71adcef491bd7ab4639766d95ab1a48e76861871c48313a124400",
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "d33ca0114a141d60d2d5017d0ea8b9911de7aca23ac024eee6b950814633b536",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
            "SandboxKey": "/var/run/docker/netns/d33ca0114a14",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "devsh-nuve_backend": {
                    "IPAMConfig": {},
                    "Links": null,
                    "Aliases": [
                        "devsh-nuve-backend-1",
                        "backend",
                        "62cc7a1eccd8"
                    ],
                    "NetworkID": "fe95195bbfe0f7f85b720b6e3818a3b7ca508ac9b2ef1caf03041d05afe91b94",
                    "EndpointID": "e1be8af10113e3a374440ccc77abcc8ac69f5fd50198a6c70da6b390f8cce420",
                    "Gateway": "172.25.16.1",
                    "IPAddress": "172.25.16.4",
                    "IPPrefixLen": 24,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:ac:19:10:04",
                    "DriverOpts": null
                },
                "webgateway": {
                    "IPAMConfig": {},
                    "Links": null,
                    "Aliases": [
                        "devsh-nuve-backend-1",
                        "backend",
                        "62cc7a1eccd8"
                    ],
                    "NetworkID": "13c3d6d38c5dfc8d542ef24fb4a4fd3a38fc5ce03af5a16715eb38543a662966",
                    "EndpointID": "026997ac736090359bb90d08dee13f69bb600e402803df8516a5e09ca71a562a",
                    "Gateway": "172.25.11.1",
                    "IPAddress": "172.25.11.12",
                    "IPPrefixLen": 24,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:ac:19:0b:0c",
                    "DriverOpts": null
                }
            }
        }
    }
]
```

Defect container setup with wrong network settings: 

```
[
    {
        "Id": "26cb01b26b9ad449972bef4a9680790a498f578b5ebe578013d95fdb92086313",
        "Created": "2023-07-14T06:31:10.807147334Z",
        "Path": "java",
        "Args": [
       ...
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 55776,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2023-07-14T06:31:11.439889668Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:1a7fbe9d41a71adcef491bd7ab4639766d95ab1a48e76861871c48313a124400",
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "53fd77375d25f60708507954558e6cd82666907c7218f4283d28a43cf98ce98b",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
            "SandboxKey": "/var/run/docker/netns/53fd77375d25",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "devsh-nuve_backend": {
                    "IPAMConfig": {},
                    "Links": null,
                    "Aliases": [
                        "devsh-nuve-backend-1",
                        "backend",
                        "26cb01b26b9a"
                    ],
                    "NetworkID": "25f1e91b47c8558e4f168fd170959b1c8d3b3d72fe9d6c1ddcfa709b235d4b93",
                    "EndpointID": "2e9b20172e6a061f880edda222053f679a5cee572749fe44cd041f545c4fbdce",
                    "Gateway": "172.25.10.1",
                    "IPAddress": "172.25.10.4",
                    "IPPrefixLen": 24,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:ac:19:0a:04",
                    "DriverOpts": null
                }
            }
        }
    }
]
```

As you can see both network settings just differ in the network "webgateway" which is correctly used one time and is missing in the other case.

Both containers are started with the same docker-compose file and the same docker-compose (v2) command. Sometimes it works, sometimes not.

I believe this issue is a concurrency issue of the Docker socket. The problem seems to disappear if i run docker-compose  with parameter "--parallel 1". Per default docker-compose is starting all services in parallel which seems to cause inconsistencies in the spawned container configurations.

### Reproduce

1. Run docker-compose up -d with a compose file with multiple services (e.g. 5) and multiple networks (e.g. 2), including one external network
2. Inspect the spawned containers and check if all containers are registered in all configured networks.
3. If yes, run docker-compose down and restart from 1.
4. For me after two or three tries some networks were missing on some containers, randomly.

### Expected behavior

Rerunning the same docker-compose command with the same compose file should result in the same network configuration of all containers.

### docker version

```bash
Client: Docker Engine - Community
 Version:           24.0.2
 API version:       1.43
 Go version:        go1.20.4
 Git commit:        cb74dfcd85
 Built:             Thu May 25 20:26:27 2023
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.21.1 (114176)
 Engine:
  Version:          24.0.2
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.4
  Git commit:       659604f
  Built:            Thu May 25 21:50:59 2023
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    24.0.2
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.0
    Path:     /Users/mberndt/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.19.1
    Path:     /Users/mberndt/.docker/cli-plugins/docker-compose
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     /Users/mberndt/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.20
    Path:     /Users/mberndt/.docker/cli-plugins/docker-extension
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v0.1.0-beta.6
    Path:     /Users/mberndt/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/mberndt/.docker/cli-plugins/docker-sbom
  scan: Docker Scan (Docker Inc.)
    Version:  v0.26.0
    Path:     /Users/mberndt/.docker/cli-plugins/docker-scan
  scout: Command line tool for Docker Scout (Docker Inc.)
    Version:  0.16.1
    Path:     /Users/mberndt/.docker/cli-plugins/docker-scout

Server:
 Containers: 18
  Running: 15
  Paused: 0
  Stopped: 3
 Images: 253
 Server Version: 24.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.49-linuxkit-pr
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 6
 Total Memory: 15.61GiB
 Name: docker-desktop
 ID: 5e57e01b-dc9e-4e49-81d9-bbca50b069b7
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Registry Mirrors:
  https://nexus-nb.data-experts.net:8444/
 Live Restore Enabled: false
 Default Address Pools:
   Base: 172.25.0.0/16, Size: 24
```

### Diagnostics ID

453E99E7-EE12-41EE-BA00-B43B3CFA8F6E/20230714065638

**Comment by Persi** [comment]
I believe this issue could be fixed by https://github.com/docker/compose/issues/10777.

**Comment by gbalduzzi** [comment]
I noticed the same issue, docker compose v2.20 should fix this but it's not shipped yet in docker for mac

**Comment by utrotzek** [comment]
> I noticed the same issue, docker compose v2.20 should fix this but it's not shipped yet in docker for mac

you can manually install a newer version from source as long this is not officially shipped: https://docs.docker.com/compose/install/linux/#install-the-plugin-manually

**Comment by hubertgrzeskowiak** [comment]
Looks like it shipped recently. I am on `Docker Compose version v2.20.2-desktop.1` now

**Comment by milas** [comment]
This was fixed in a newer Docker Desktop / Compose release. Please upgrade and create a new issue if you still have problems. Thanks for the report!
